### ISSUE TITLE
Reenable Link tests from Pybind PR, add Link type conversions

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#140]
after #136  

https://github.com/htm-community/nupic.cpp/pull/136#issuecomment-443524149
> Now...the current implementation does not allow array data type conversions within the Links between Regions. It will require that the Output type of the sending Region must have the SAME type as the receiving Region's Input. One of the future PR's will be to provide some portion of automatic data type conversion within the Link. 

- [ ] add conversions
- [ ] reenable Link tests

**Comment by breznak** [comment]
@dkeeney How come it does work now, and doesn't with Pybind? Is PyB more strict with the conversions?

**Comment by dkeeney** [comment]
No, the problem is that the output type of a sending region MUST match the input type of a receiving region.  (nothing to do with PyBind).  So a with the advent of the issparse flag the link must not only check the type but also the issparse flag.   What I was mumbling about is that WHEN we add conversions there will be another conversion to make and I cannot just switch on BasicType.  

I don't know why the issparse flag killed the LinkTest other than perhaps the merge from numenta missed changing the Specs in the regions.  We need to go investigate what all they did in numenta's repository.  

There is more involved in this Link thing.  Each region Spec also has a singleNodeOnly boolean.  If true, then the Link does not support dimensions.  In other words, the LinkPolicy cannot be used to slice and dice a buffer based on dimensions within the Link as data flows between regions. What I noticed was that most if not all regions in the Python code have this flag as true.  In fact the only place I find it false is in the TestNode.  What that means is that we could actually remove LinkPolicy.hpp, LinkPolicyFactory.cpp/hpp, TestFanIn2LinkPolicy.cpp/hpp, UniformLinkPolicy.cpp/hpp, and remove a lot of complication from the Link object without affecting anything.  After doing that we could then add in the link data conversions without having to worry about conflicts with LinkPolicy.  

 I added the array type conversion functions to ArrayBase and BasicType with the intension of eventually allowing array type conversions within the Link.   I coded that all up (removed LinkPolicy and dimensions, added conversions) and it worked very well but it borders on changing the API. Technically someone could specify a link policy on a Link but it would be ignored; the same as if the singleNodeOnly  flag was on.  So the effect would be the same as the API.

I guess I should take some time to go check what numenta has done lately in their repository.....but not until we get PyBind working.  Back to work.
